File: leetcode95_Recursion.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func generateTrees(n int) []*TreeNode {
    if n == 0 {
        return []*TreeNode{}
    }
    return genTrees(1,n)
}

func genTrees(start int, end int) []*TreeNode {
    list := make([]*TreeNode, 0)
    if start > end {
        var n *TreeNode
        list = append(list, n)
        return list
    }
    
    var left []*TreeNode
    var right []*TreeNode
    for i:= start; i <= end; i++ {
        left= genTrees(start, i - 1)
        right= genTrees(i+1, end)
    
        for _, lNode := range left {
            for _, rNode := range right {
                root:= new(TreeNode)
                root.Val = i
                root.Left = lNode
                root.Right = rNode
                list = append(list, root)
            }
        }
    }
    return list
}

File: leetcode95_Recursion.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        if(n == 0)
            return new ArrayList<TreeNode>();
        return generate(1, n);
    }
    public List<TreeNode> generate(int left, int right){
        List<TreeNode> res = new ArrayList<>();
        if(left > right){
            res.add(null);
            return res;
        }
        for(int i = left; i <= right; i++){
            List<TreeNode> lefts = generate(left, i - 1);
            List<TreeNode> rights = generate(i + 1, right);
            for(int j = 0; j < lefts.size(); j++){
                for(int k = 0; k < rights.size(); k++){
                    TreeNode root = new TreeNode(i);
                    root.left = lefts.get(j);
                    root.right = rights.get(k);
                    res.add(root);
                }
            }
        }
        return res;
    }
}
